### PR TITLE
Performance: Fetch permissions for visible patterns only

### LINKED ISSUE #64219: Performance Degradation in the Block Editor after version 18.0.1 for instances with 3000+ block patterns
### Description

I was able to detect significant performance degradation after inserting 3000 block patterns in a clean install of WordPress 6.6. I tested different versions of the Gutenberg plugin, and it looks like this behavior was not present before (and including) version 18.0.1 and was likely introduced after (and including) version 18.1.0 of Gutenberg.

There are delays in multiple actions in the editor, and they all seem to be JS-related, the post editor page loads with no problems, but, after running all the `wp-json/wp/v2/blocks?context=edit&per_page=100&_locale=user&page=<index>` REST requests to load the block patterns in the background the slowness begins - for example: clicking/focusing on a block that is being edited takes around 1 second.

I ran some performance tests on Chrome. I have a hard time figuring out what can be wrong there though - this is the time it takes to focus on a button and for the top context bar to show up:
image

### Step-by-step reproduction instructions

1. Create 3000+ reusable blocks with the CLI using `wp eval 'for ($i = 1; $i <= 3000; $i++) { wp_insert_post(["post_title" => "Reusable Block " . $i, "post_content" => "<!-- wp:paragraph --><p>This is block " . $i . "</p><!-- /wp:paragraph -->", "post_status" => "publish", "post_type" => "wp_block"]); }'`
2. Open the post editor, create two button blocks with some text in it
3. After all the `wp-json/wp/v2/blocks?context=edit&per_page=100&_locale=user&page=<index>` REST requests run in the background (for me it takes around 10 seconds) the editor lag should begin. Click on one button, then click on another button. The context bar for the button should take a long time to show up, more than 1 second.

### Screenshots, screen recording, code snippet

https://github.com/user-attachments/assets/67ea9b58-3754-4036-94e8-69cef61a419d

### Environment info

- WordPress 6.6
- Gutenberg plugin 18.1.0
- Google Chrome (127.0.6533.89)
- macOS Sonoma 14.6
- [VIP Local Development Environment](https://docs.wpvip.com/vip-local-development-environment/)

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### PR DESCRIPTION
**Pull request by youknowriad**
## What?

When rendering patterns, we need the user permissions. Unfortunately right now, the only way to fetch the permissions is to do a request per pattern. There's some ongoing work to improve that but in the mean time, this PR restores the previous behavior where we fetch the permissions for visible patterns only.

## Testing Instructions

Please check #64219 for detailed steps.

### CODE REVIEW

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the quick work here @youknowriad 🚀 

This is testing well for me and makes a huge difference compared to trunk. 

Using the videos below to roughly check the time it took to create a new pattern and navigate to the pattern editor, each branch took:

- Trunk: about 2min 22secs
- This PR: < 1s

Similar improvements are also seen for other tasks such as deleting a pattern. 

When initially loading the patterns page with a thousand patterns, there's still a considerable delay before the My Patterns category and count shows. That seems like a separate opportunity for fine-tuning though.

It's been a while since I've looked at the Patterns page, so feel free to wait for a second approval but I'm happy to give this a ✅  now.

| Trunk | This PR |
|---|---|
| <video src="https://github.com/user-attachments/assets/4592cf8c-9d12-433a-80b5-1d4ef86d5651" /> | <video src="https://github.com/user-attachments/assets/8e9335bc-770f-4ba2-bfde-73c77f73a008" /> |

**Comment by Mamaduka** [comment]
The #64504 should also improve permissions requests.

**Pull request merged**
